### PR TITLE
add universal consistency middleware

### PR DESCRIPTION
**Pull request by ecordell**
Depends on #392

### CODE REVIEW

File: internal/services/v1alpha1/schema_test.go [hunk 1]
```diff
@@ -28,8 +28,27 @@ func TestSchemaReadNoPrefix(t *testing.T) {
 	ds, err := memdb.NewMemdbDatastore(0, 0, memdb.DisableGC, 0)
 	require.NoError(t, err)
 
-	srv := NewSchemaServer(ds, PrefixRequired)
-	_, err = srv.ReadSchema(context.Background(), &v1alpha1.ReadSchemaRequest{
+	// test relies on middleware
```
Comment (jakedt) [inline internal/services/v1alpha1/schema_test.go:31]: There are far too many copies of code that looks just like this in this PR. Why?

File: pkg/testutil/require.go [hunk 2]
```diff
@@ -3,30 +3,22 @@
...
+	v0 "github.com/authzed/authzed-go/proto/authzed/api/v0"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
 	"github.com/stretchr/testify/require"
 )
 
 // RequireEqualEmptyNil is a version of require.Equal, but considers nil
 // slices/maps to be equal to empty slices/maps.
 func RequireEqualEmptyNil(t *testing.T, expected, actual interface{}, msgAndArgs ...interface{}) {
-	expectedVal := reflect.ValueOf(expected)
```
Comment (jakedt) [inline pkg/testutil/require.go:15]: These changes should probably also be in a separate commit.

File: internal/services/v0/acl.go [hunk 3]
```diff
@@ -258,20 +239,13 @@ func (as *aclServer) Read(ctx context.Context, req *v0.ReadRequest) (*v0.ReadRes
 }
 
 func (as *aclServer) Check(ctx context.Context, req *v0.CheckRequest) (*v0.CheckResponse, error) {
-	atRevision, err := as.pickBestRevision(ctx, req.AtRevision)
```
Comment (jakedt) [inline internal/services/v0/acl.go:261]: It's a little unclear why `Check` and `ContentChangeCheck` will behave differently at first glance. I wonder if we could improve the ergonomics at the call site of `MustRevisionFromContext`.

**Review by jakedt — CHANGES_REQUESTED** [review]

File: internal/services/v0/acl.go (hunk 3 above)
Comment (ecordell) [reply internal/services/v0/acl.go:261]: Is it worth doing much more than a comment here, since it's the v0 api? (I think the consistency block logic is much clearer in v1).

File: internal/services/v1alpha1/schema_test.go (hunk 1 above)
Comment (ecordell) [reply internal/services/v1alpha1/schema_test.go:31]: I was planning to use to buffconn PR to clean these up later, but I'll do it now so it's more reviewable.

File: internal/services/v0/acl.go (hunk 3 above)
Comment (ecordell) [reply internal/services/v0/acl.go:261]: (added a comment to ContentChangeCheck)

File: pkg/testutil/require.go (hunk 2 above)
Comment (ecordell) [reply pkg/testutil/require.go:15]: I didn't track down exactly why this started happening, but after adding the consistency middleware some of the tests that assert with `RequireEqualEmptyNil` had some deeply nested types be unequal (because one side was nil and the other was an empty slice). I thought it was easier to make the test helper (which already collapsed those cases for top-level values) work for nested types too (and also print out more helpful error messages when they fail) instead of touching every test expectation.

TL;DR this is pretty coupled to the other changes in this commit

Comment (jakedt) [reply pkg/testutil/require.go:15]: That's fine, but can't you just put it in a commit before you introduce the universal consistency? universal consistency has this as a pre-requisite IIUC.

File: cmd/spicedb/servetesting_integration_test.go [hunk 4]
```diff
@@ -159,13 +160,14 @@ func newTester(t *testing.T, containerOpts *dockertest.RunOptions) (*spicedbHand
 		}
 	}
 
+	conn, err := grpc.Dial(fmt.Sprintf("localhost:%s", port),
```
[the remark this answers is not in the document]
Comment (jakedt) [reply cmd/spicedb/servetesting_integration_test.go]: Did you move this to a separate PR or something?

[the remark this answers is not in the document]
Comment (ecordell) [reply cmd/spicedb/servetesting_integration_test.go]: I just removed it since it was unrelated. I'll have to touch this file when I update the test server anyway (so the linter will complain that I'm using a deprecated api and I'll have to change it there)

File: internal/services/v0/namespace_test.go [hunk 5]
```diff
@@ -111,29 +90,30 @@ func TestNamespaceChanged(t *testing.T) {
...
-			ns.Namespace(
-				"folder",
-				ns.Relation("anotherrel",
-					nil,
-					ns.AllowedRelation("folder", "..."),
-				),
-			),
-			"cannot delete relation `viewer` in definition `folder`, as a relationship references it",
-		},
+		// TODO: review this test
```
Comment (jakedt) [inline internal/services/v0/namespace_test.go:93]: Review it for what? Was it reviewed? Is it just failing?

**Review by jakedt — CHANGES_REQUESTED** [review]

File: internal/services/v0/namespace_test.go (hunk 5 above)
Comment (ecordell) [reply internal/services/v0/namespace_test.go]: Fixed - the tuple being written wasn't valid under the namespace config; the previous test wasn't type-checking
